packages/relay/src/lib/services/hbarLimitService/index.ts

```diff
@@ -322,13 +328,28 @@ export class HbarLimitService implements IHbarLimitService {
   }
 
   /**
-   * Gets a new reset timestamp for the rate limiter.
-   * @returns {Date} - The new reset timestamp.
-   * @private
+   * Calculates the next reset timestamp for the rate limiter.
+   *
+   * This method determines the next reset timestamp based on the current reset timestamp
+   * and the limit duration. If the current reset timestamp is not defined, it initializes
+   * the reset timestamp to midnight of the current day. It then iteratively adds the limit
+   * duration to the reset timestamp until it is in the future.
+   *
+   * @returns {Date} - The next reset timestamp.
    */
   private getResetTimestamp(): Date {
-    const tomorrow = new Date(Date.now() + HbarLimitService.ONE_DAY_IN_MILLIS);
-    return new Date(tomorrow.setHours(0, 0, 0, 0));
+    const todayAtMidnight = new Date().setHours(0, 0, 0, 0);
+
+    let resetDate = this.reset ? new Date(this.reset.getTime()) : new Date(todayAtMidnight);
+    if (resetDate.getTime() < Date.now()) {
+      // 1. Calculate the difference between the current time and the reset time.
+      // 2. Determine how many intervals of size `limitDuration` have passed since the last reset.
+      // 3. Calculate the new reset date by adding the required intervals to the original reset date.
+      const intervalsPassed = Math.ceil((Date.now() - resetDate.getTime()) / this.limitDuration);
+      resetDate = new Date(resetDate.getTime() + intervalsPassed * this.limitDuration);
+    }
+
+    return resetDate;
```

Comment:
Let's say on the first startup of the Relay, the `resetDate` is set to midnight of the current date. For simplicity, let's assign it a UNIX timestamp of 18000, and assume `limitDuration = 80 seconds`.

After midnight, a new request comes in at `Date.now = 18080` (80 seconds after midnight). The calculation would be:

`intervalsPassed = (18080 - 18000) / 80 = 1 interval passed`.

Then, the new `resetDate` would be:

`resetDate = 18000 + 1 * 80 = 18080` (which is the current `Date.now`). 

In this case, why don't we simply set `resetDate = Date.now` instead of calculating the number of intervals passed?

In another scenario, assuming midnight is still at 18000, if a request comes in at `18100` (100 seconds after midnight), the calculation would be:

`intervalsPassed = Math.ceil(100 / 80) = 2 intervals (or 160 seconds)`.

Thus, the new `resetDate` would be:

`resetDate = 18000 + 2 * 80 = 18160`.

This creates a gap of 60 seconds between when the request was made and the reset time. 

I guess I am confused with this logic of calculating the resetDate.

Reply:
Aside from all that above, the `getResetTimestamp()` method is used to retrieve the `resetDate` and reset `this.reset` in the `resetLimiter()` method. This part is also somewhat confusing.

Using the same example, let's assume at startup, `getResetTimestamp()` returns a `resetDate` set to midnight (we'll use 18000 to remain consistent with the previous example). This makes sense because `this.reset = 18000`, and it’s in the future compare to startup time.

Now, let’s say a request comes in after midnight at `18080` (80 seconds after midnight). At this point, `shouldResetLimiter()` is triggered and compares `Date.now()`, which is `18080`, to `this.reset`, which is `18000`. Since `Date.now()` is greater than `this.reset`, `shouldResetLimiter()` returns `true`, which makes sense so far. 

When `shouldResetLimiter()` returns `true`, `resetLimiter()` is called, and `this.reset` is set to `this.getResetTimestamp()`. So far still makes perfect sense.

However, with the current logic in `getResetTimestamp()`, when the request arrives at `18080`, `this.reset` is still at `18000`. Therefore, it falls into the following block:
```typescript
      const intervalsPassed = Math.ceil((Date.now() - resetDate.getTime()) / this.limitDuration);
      resetDate = new Date(resetDate.getTime() + intervalsPassed * this.limitDuration);
```
This calculates the `resetDate` as `18080`, which is the same time as when the request was made. This means the `resetDate` will always be in the past from this point onward, because any future requests will come after the `18080` timestamp. As a result, `shouldResetLimiter()` will always return `true`, since `Date.now()` will always be greater than the reset timestamp, which remains at `18080`.

This doesn’t seem logical, as the `resetDate` should always be in the future and at least 24 hours (1 day) apart from the previous reset timestamp.

Reply:
> This calculates the `resetDate` as `18080`, which is the same time as when the request was made. This means the `resetDate` will always be in the past from this point onward, because any future requests will come after the `18080` timestamp. As a result, `shouldResetLimiter()` will always return `true`, since `Date.now()` will always be greater than the reset timestamp, which remains at `18080`.

1. this is not really possible to happen, it's an extreme edge case
2. this is not true, because what would happen is that on the next request `shouldResetLimiter()` will return true and `getResetTimestamp()` will be called again by `resetLimiter()` to calculate the next reset timestamp, which will be `18160`

Reply:
> Thus, the new `resetDate` would be:
> 
> `resetDate = 18000 + 2 * 80 = 18160.`
> 
> This creates a gap of 60 seconds between when the request was made and the reset time.

There is no "gap" because if `getResetTimestamp()` was called, this means that `resetLimiter()` was called and the hbar limiter was resetted.

Reply:
Overall think of `getResetTimestamp()` as a method which is called to give you the **_next reset timestamp_** and we call it only when we have resetted the limiter.

Reply:
Discussed offline

Reply:
Was a confusion I had behind the idea of the limitDuration carrying from the old implementation. Fully resolved offline.